### ISSUE TITLE
WebApplicationFactory does not work from when CreateHostBuilder is from an inherited class, instead of the entry point assembly.

### ISSUE THREAD

**Issue by PureKrome** [aspnet/AspNetCore#17994]
### Related

This issue is related to #8210 and (archived) [Hosting #1517](https://github.com/aspnet/Hosting/issues/1517)

---
## Problem

The `WebApplicationFactory` doesn't work right when the "entry point assembly" references _another_ assembly which contains the `CreateHostBuilder` logic. Current code (I think) requires a method called `CreateHostBuilder` to exist in the entry point assembly, only.

---

## Repo

A full repo of this problem [can be found here](https://github.com/PureKrome/WebAppFactoryRepo) in GitHub.

---
## Description

I have a pretty standard ASP.NET Core 3.1 web api application and i'm trying to use the `WebAppFactory` to test my web application.

For my `program.cs` file, instead of adding my custom code in there (e.g. wire up Serilog, etc), I actually have a nuget package which I then re-use everywhere. So if I change my `program.cs` logic, I can just update various webapi applications with the latest nuget, instead of having to change the copy/paste code in each webapi `program.cs` class. Works really well.

Given this type of setup, the `WebApplicationFactory` fails to `ResolveHostBuilderFactory` because the _entry point assembly_ doesn't contain the `CreateHostBuilder<T>` method. And therefore stuff isn't working 100% right.

For example.

this is my `program.cs`

```
public class Program
{
    public static Task Main()
    {
        var options = new MainOptions
        {
            FirstLoggingInformationMessage = "~~~ Starting FooBar Web Api  ~~~"
        };

        return Homely.AspNetCore.Hosting.CoreApp.Program.Main<Startup>(options);
    }
}
```

That's it. Notice how it's very different to the default template _on first glance_. So my `program.cs` class literally says: -> call some other assembly's `Main` method and do stuff. It's this _other_ `Main<T>(..)` method that contains the typical template logic. 

([That `Main` method logic](https://github.com/Homely/Homely.AspNetCore.Hosting.CoreApp/blob/master/src/Homely.AspNetCore.Hosting.CoreApp/Program.cs#L57) is literally -> setup Serilog, do the normal program.cs templated stuff ... then flush the Serilog on exit. So just logging plumbing wrapped around the _default_ program.cs Main logic).

Okay - so my entry point assembly just calls some _other_ assembly. Nothing crazy.

Except ... it doesn't work in this case because the `WebApplicationFactory.CreateHostBuilder()` [code does this](https://github.com/aspnet/AspNetCore/blob/f7f80fdbaaa0283e8e2e89ceed7abf7fe413b17c/src/Mvc/Mvc.Testing/src/WebApplicationFactory.cs#L318)....

```
protected virtual IHostBuilder CreateHostBuilder()
{
    var hostBuilder = HostFactoryResolver.ResolveHostBuilderFactory<IHostBuilder>(typeof(TEntryPoint).Assembly)?.Invoke(Array.Empty<string>());
    
<snip>
}
```

and [that `RHBF` method then does this](https://github.com/aspnet/Extensions/blob/master/src/Shared/src/HostFactoryResolver/HostFactoryResolver.cs#L32) ...

```
        public static Func<string[], THostBuilder> ResolveHostBuilderFactory<THostBuilder>(Assembly assembly)
        {
            return ResolveFactory<THostBuilder>(assembly, CreateHostBuilder);
        }

        private static Func<string[], T> ResolveFactory<T>(Assembly assembly, string name)
        {
            var programType = assembly?.EntryPoint?.DeclaringType;
            if (programType == null)
            {
                return null;
            }

           <snip>
        }
```

and the `assembly?.EntryPoint?.DeclaringType` ends up returning `null`. Sadness :(

So .. I'm not sure what can be done here, for this scenario.
It's like I wish the code could 'see' that the entry point assembly is not the right place to check, but _another_ assembly which I tell it, is what should be checked.

Secondly, I also tried this:

```
public class WebAppFactory : WebApplicationFactory<Homely.AspNetCore.Hosting.CoreApp.Program>
```

but that errored with: `System.InvalidOperationException : The provided Type 'Program' does not belong to an assembly with an entry point. A common cause for this error is providing a Type from a class library.` That's _exactly_ what I'm trying to do, though :P

So - would love some help, please?

---

```
$ dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   3.1.100
 Commit:    cd82f021f4

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18362
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.100\

Host (useful for support):
  Version: 3.1.0
  Commit:  65f04fb6db

```

**Comment by DamianEdwards** [comment]
@davidfowl is this fixed by your recent changes to support top-level statements, etc.?

**Comment by davidfowl** [comment]
Yes it'll work but @PureKrome your repo would need to override the right method `ConfigureWebHost` instaed of `CreateHostBuilder`

**Comment by PureKrome** [comment]
> Yes it'll work but @PureKrome your repo would need to override the right method `ConfigureWebHost` instaed of `CreateHostBuilder`

Thanks @davidfowl for the reply! The "Yes it'll work" ... is this in any specific version of .NET Core? Or in .NET 6?

**Comment by davidfowl** [comment]
.NET 6

**Comment by javiercn** [comment]
@davidfowl seems this can be closed?

**Comment by ChrisRichner** [comment]
Hi All

@PureKrome  I'm currently trying to extract the Startup classes in a Shared Class Library (Nuget) as well.  Facing the extact same Error in .net 7.0.

Did you find a working solution to this in the meanwhile?

Thanks

**Comment by PureKrome** [comment]
@ChrisRichner nope. I don't _think_ I did. I've also started moving over to minimal api's .. so _again_ i'm not sure how this will play out. My issue was that I like to use some _common_ startup stuff, which is why I put it into a shared class library / nuget ... which sounds like what you were/are doing.

I've yet to do this with minimal api's and "see" how that would look ... let along trying any E2E testing using a WAF with minimal api's.

I'm sorry I'm not the messenger of positive news :(

**Comment by ChrisRichner** [comment]
@PureKrome  Thank you for your quick Feedback on this Issue. 

I've found a working solution where I only need to "fix" some lines of code to make it work

    protected override IWebHostBuilder CreateWebHostBuilder()
    {  
        // Just referencing my Head Project Program.cs. CreateWebHostBuilder() which then just calls the SharedLibrary.CreateWebHostBuilder() internally.
        // Hope this does the trick for my current Project  (fingers crossed :)
        var builder = Program.CreateWebHostBuilder(Array.Empty<string>(), new StartupOptions());

        if (builder is not null)
        {
            return builder.UseEnvironment(Environments.Development);
        }

        return null;
    }

**Comment by PureKrome** [comment]
@ChrisRichner interesting! how does this work?
